### PR TITLE
[MLIR] RemoveDeadValues: Allowing IRs with global constants to get dead values removed

### PR DESCRIPTION
**Pull request by parsifal-47**
This change is related to discussion:
https://discourse.llvm.org/t/question-on-criteria-for-acceptable-ir-in-removedeadvaluespass/83131

I do not know the original reason to disallow the optimization on modules with global private constant. Please let me know what am I missing, I will be happy to make it better. Thank you!

CC: @Wheest

### CODE REVIEW

File: mlir/lib/Transforms/RemoveDeadValues.cpp
Comment (Wheest) [inline mlir/lib/Transforms/RemoveDeadValues.cpp:583]: So the original comment above says:

```cpp
  // The removal of non-live values is performed iff there are no branch ops,
  // all symbol ops present in the IR are function-like, and all symbol user ops
  // present in the IR are call-like.
```

and the error message includes `non-call symbol user ops`.  With this PR I believe that IR with non-call symbol user ops are no longer excluded?

Perhaps the error message/comment should be updated to reflect that.

File: mlir/test/Transforms/remove-dead-values.mlir
Comment (Wheest) [inline mlir/test/Transforms/remove-dead-values.mlir:6]: is `@__dont_touch_unacceptable_ir` still a meaningful name here, since we expect the IR to be updated?

**Review by Wheest — COMMENTED** [review]
From my perspective it should be fine if IR with named symbols like constants use this pass (my recent PR did this for named modules #109990).

A justification for keeping the existing behaviour could be: "that which has a name should be handled more carefully, and not removed unless specifically requested."

However, since this is a blocker for some of my own work, I don't support it, and think this PR should go through.

That being said, I could imagine adding this behaviour back in with `--remove-dead-values="remove-user-symbols"`

**Comment by River707** [comment]
From my perspective I'd love to see all of the `acceptableIR` checks removed completely. This passes presents itself as some general aggressive DCE, but is currently only applicable in a very limited number of situations. I'd support any form of additions that move it closer towards a more general pass.

File: mlir/test/Transforms/remove-dead-values.mlir
Comment (parsifal-47) [reply mlir/test/Transforms/remove-dead-values.mlir]: thank you, done!

File: mlir/lib/Transforms/RemoveDeadValues.cpp
Comment (parsifal-47) [reply mlir/lib/Transforms/RemoveDeadValues.cpp]: updated both the comment and error message, thank you!

**Comment by codemzs** [comment]
@River707 @Wheest Who is the right person to ask to commit this change?

**Comment by codemzs** [comment]
Hello joker-eph,

It appears that you have recently reviewed and committed changes to this file. Would it be possible for you to review this change?

File: mlir/lib/Transforms/RemoveDeadValues.cpp
Comment (joker-eph) [inline mlir/lib/Transforms/RemoveDeadValues.cpp:581]: This last check is also not clear to me. But I'm not sure if this has all to do with some aspects of the dataflow framework?

**Review by joker-eph — APPROVED** [review]
It's not clear why this check is in-place (unfortunately the comment documents the "what" and not the "why"...).

In any case we could tighten this back with motivating examples, if they show up, and I suspect the checks could be more targeted: for example instead of disabling the pass entirely, we could disable it on a per-function basis.

**Comment by codemzs** [comment]
> It's not clear why this check is in-place (unfortunately the comment documents the "what" and not the "why"...).
> 
> In any case we could tighten this back with motivating examples, if they show up, and I suspect the checks could be more targeted: for example instead of disabling the pass entirely, we could disable it on a per-function basis.

@joker-eph I was curious if there is anything else needed for this change to be committed?

**Comment by joker-eph** [comment]
It is approved, and the author has commit access, so I didn't feel like committing in their place. @parsifal-47 what's up?

**Comment by parsifal-47** [comment]
hmm, I see the following:
"This branch has no conflicts with the base branch
Only those with [write access](https://docs.github.com/articles/what-are-the-different-access-permissions) to this repository can merge pull requests."

**Comment by parsifal-47** [comment]
sorry, maybe I do not understand the workflow, the PR is approved, should I look for motivating examples?
I was thinking that we are waiting for all potentially interested people to see it before it gets merged.

**Comment by codemzs** [comment]
> It is approved, and the author has commit access, so I didn't feel like committing in their place. @parsifal-47 what's up?

@joker-eph Based on my offline conversation with @parsifal-47 it appears he lacks write access. Would it be possible for you to commit and merge this change into `main`? Thank you.

**Pull request merged**
